File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/submodel/metamodel/map/submodelelement/SubmodelElement.java
package org.eclipse.basyx.submodel.metamodel.map.submodelelement;

import java.util.Collection;
import java.util.Map;

import org.eclipse.basyx.submodel.metamodel.api.dataspecification.IEmbeddedDataSpecification;
import org.eclipse.basyx.submodel.metamodel.api.qualifier.haskind.ModelingKind;
import org.eclipse.basyx.submodel.metamodel.api.qualifier.qualifiable.IConstraint;
import org.eclipse.basyx.submodel.metamodel.api.reference.IReference;
import org.eclipse.basyx.submodel.metamodel.api.reference.enums.KeyElements;
import org.eclipse.basyx.submodel.metamodel.api.submodelelement.ISubmodelElement;
import org.eclipse.basyx.submodel.metamodel.map.modeltype.ModelType;
import org.eclipse.basyx.submodel.metamodel.map.qualifier.HasDataSpecification;
import org.eclipse.basyx.submodel.metamodel.map.qualifier.HasSemantics;
import org.eclipse.basyx.submodel.metamodel.map.qualifier.LangStrings;
import org.eclipse.basyx.submodel.metamodel.map.qualifier.Referable;
import org.eclipse.basyx.submodel.metamodel.map.qualifier.haskind.HasKind;
import org.eclipse.basyx.submodel.metamodel.map.qualifier.qualifiable.Qualifiable;
import org.eclipse.basyx.vab.model.VABModelMap;


/**
 * SubmodelElement as defined in "Details of the Asset Administration Shell"
 * 
 * @author schnicke
 *
 */
public class SubmodelElement extends VABModelMap<Object> implements ISubmodelElement {
	public static final String MODELTYPE = "SubmodelElement";

	protected SubmodelElement() {
		// Add model type
		putAll(new ModelType(MODELTYPE));

		putAll(new HasDataSpecification());
		putAll(new Referable());
		putAll(new Qualifiable());
		putAll(new HasSemantics());
		putAll(new HasKind());
	}
	
	/**
	 * Creates a SubmodelElement object from a map
	 * 
	 * @param obj a SubmodelElement object as raw map
	 * @return a SubmodelElement object, that behaves like a facade for the given map
	 */
	public static SubmodelElement createAsFacade(Map<String, Object> obj) {
		SubmodelElement ret = new SubmodelElement();
		ret.setMap(obj);
		return ret;
	}

	@Override
	public Collection<IReference> getDataSpecificationReferences() {
		return HasDataSpecification.createAsFacade(this).getDataSpecificationReferences();
	}

	public void setDataSpecificationReferences(Collection<IReference> ref) {
		HasDataSpecification.createAsFacade(this).setDataSpecificationReferences(ref);
	}

	@Override
	public Collection<IEmbeddedDataSpecification> getEmbeddedDataSpecifications() {
		return HasDataSpecification.createAsFacade(this).getEmbeddedDataSpecifications();
	}

	public void setEmbeddedDataSpecifications(Collection<IEmbeddedDataSpecification> embeddedDataSpecifications) {
		HasDataSpecification.createAsFacade(this).setEmbeddedDataSpecifications(embeddedDataSpecifications);
	}

	@Override
	public String getIdShort() {
		return Referable.createAsFacade(this, getKeyElement()).getIdShort();
	}

	protected KeyElements getKeyElement() {
		return KeyElements.SUBMODELELEMENT;
	}

	@Override
	public String getCategory() {
		return Referable.createAsFacade(this, getKeyElement()).getCategory();
	}

	@Override
	public LangStrings getDescription() {
		return Referable.createAsFacade(this, getKeyElement()).getDescription();
	}

	@Override
	public IReference getParent() {
		return Referable.createAsFacade(this, getKeyElement()).getParent();
	}

	public void setIdShort(String idShort) {
		Referable.createAsFacade(this, getKeyElement()).setIdShort(idShort);
	}

	public void setCategory(String category) {
		Referable.createAsFacade(this, getKeyElement()).setCategory(category);
	}

	public void setDescription(LangStrings description) {
		Referable.createAsFacade(this, getKeyElement()).setDescription(description);
	}

	public void setParent(IReference obj) {
		Referable.createAsFacade(this, getKeyElement()).setParent(obj);
	}

	public void setQualifiers(Collection<IConstraint> qualifiers) {
		Qualifiable.createAsFacade(this).setQualifiers(qualifiers);
	}

	@Override
	public Collection<IConstraint> getQualifiers() {
		return Qualifiable.createAsFacade(this).getQualifiers();
	}

	@Override
	public IReference getSemanticId() {
		return HasSemantics.createAsFacade(this).getSemanticId();
	}

	public void setSemanticID(IReference ref) {
		HasSemantics.createAsFacade(this).setSemanticID(ref);
	}

	@Override
	public ModelingKind getModelingKind() {
		return HasKind.createAsFacade(this).getModelingKind();
	}

	public void setModelingKind(ModelingKind kind) {
		HasKind.createAsFacade(this).setModelingKind(kind);
	}

	@Override
	@SuppressWarnings("unchecked")
	public String getModelType() {
		return (String) ((Map<String, Object>) get(ModelType.MODELTYPE)).get(ModelType.NAME);
	}

	@Override
	public IReference getReference() {
		return Referable.createAsFacade(this, getKeyElement()).getReference();
	}
}

File: sdks/c++/basys.sdk.cc/src/submodel/submodel/simple/submodelelement/SubmodelElement.cpp
#include <BaSyx/submodel/simple/submodelelement/SubmodelElement.h>

using namespace basyx::submodel;
using namespace basyx::submodel::simple;
using namespace basyx::submodel::api;

SubmodelElement::SubmodelElement(const std::string & idShort, ModelingKind kind)
	: referable(idShort)
	, kind(kind)
{

}

const IReference & SubmodelElement::getSemanticId() const
{
	return this->semanticId;
};

void SubmodelElement::setSemanticId(const api::IReference & semanticId)
{
	this->semanticId = semanticId;
};

void SubmodelElement::addDataSpecification(const Reference & reference)
{
	this->dataSpecification.addDataSpecification(reference);
}

const std::vector<Reference> SubmodelElement::getDataSpecificationReference() const
{
	return this->dataSpecification.getDataSpecificationReference();
}

const std::string & SubmodelElement::getIdShort() const
{
	return this->referable.getIdShort();
}

void SubmodelElement::setCategory(const std::string &category)
{
  this->referable.setCategory(category);
}

const std::string * const SubmodelElement::getCategory() const
{
	return this->referable.getCategory();
}

simple::LangStringSet & SubmodelElement::getDescription()
{
	return this->referable.getDescription();
}

const simple::LangStringSet & SubmodelElement::getDescription() const
{
	return this->referable.getDescription();
}

IReferable * SubmodelElement::getParent() const
{
	return this->referable.getParent();
}

void SubmodelElement::setParent(IReferable * parent)
{
	this->setParent(parent);
};

ModelingKind SubmodelElement::getKind() const
{
	return this->kind;
}

void SubmodelElement::addFormula(const api::IFormula &formula)
{
  this->qualifiable.addFormula(formula);
}

void SubmodelElement::addQualifier(const api::IQualifier &qualifier)
{
  this->qualifiable.addQualifier(qualifier);
}

std::vector<Qualifier> SubmodelElement::getQualifiers() const
{
  return this->qualifiable.getQualifiers();
}

std::vector<Formula> SubmodelElement::getFormulas() const
{
  return this->qualifiable.getFormulas();
}

ModelTypes SubmodelElement::GetModelType() const
{
  return this->modelType;
}

Key SubmodelElement::getKey(bool local) const 
{
	return this->referable.getKey(local);
}

simple::Reference SubmodelElement::getReference() const
{
	return this->referable.getReference();
}
